File: packages/engine/src/liveEffects/paletteCore.ts
/**
 * Shared quantization core for the Dither and Palette Snap effects.
 *
 * Owns: palette representation, color-distance metrics, nearest-color lookup
 * (with a uniform-grid LUT acceleration for large palettes), and palette
 * generation (median cut + k-means refinement). The Dither and Palette Snap
 * kernels both consume these primitives — nothing is duplicated.
 *
 * Metrics operate on 0..1 components internally so gamma/space conversion is
 * explicit and testable.
 */

import { linearSrgbToOklab } from '@varve/shared';
import { seeded01, srgbToLinear01 } from './prng';

export type PaletteColor = readonly [number, number, number];

export type ColorMetric = 'rgb' | 'linear-rgb' | 'lab' | 'oklab';

const LUT_GRID = 32;

/** Compute the squared distance between an sRGB byte color and a palette color. */
export function paletteDistance(
  r: number,
  g: number,
  b: number,
  pr: number,
  pg: number,
  pb: number,
  metric: ColorMetric,
): number {
  switch (metric) {
    case 'rgb': {
      const dr = r - pr;
      const dg = g - pg;
      const db = b - pb;
      return dr * dr + dg * dg + db * db;
    }
    case 'linear-rgb': {
      const lr = srgbToLinear01(r) - srgbToLinear01(pr);
      const lg = srgbToLinear01(g) - srgbToLinear01(pg);
      const lb = srgbToLinear01(b) - srgbToLinear01(pb);
      return lr * lr + lg * lg + lb * lb;
    }
    case 'lab':
    case 'oklab': {
      const [l1, a1, b1] = toLabSpace(r, g, b, metric);
      const [l2, a2, b2] = toLabSpace(pr, pg, pb, metric);
      const dl = l1 - l2;
      const da = a1 - a2;
      const db = b1 - b2;
      return dl * dl + da * da + db * db;
    }
  }
}

function toLabSpace(
  r: number,
  g: number,
  b: number,
  metric: ColorMetric,
): [number, number, number] {
  if (metric === 'oklab') {
    return linearSrgbToOklab([srgbToLinear01(r), srgbToLinear01(g), srgbToLinear01(b)]);
  }
  // Lab via XYZ D65 (matches @varve/shared analytical conversion path).
  const rl = srgbToLinear01(r);
  const gl = srgbToLinear01(g);
  const bl = srgbToLinear01(b);
  let x = rl * 0.4124 + gl * 0.3576 + bl * 0.1805;
  const y = rl * 0.2126 + gl * 0.7152 + bl * 0.0722;
  let z = rl * 0.0193 + gl * 0.1192 + bl * 0.9505;
  x /= 0.95047;
  z /= 1.08883;
  const fx = fLab(x);
  const fy = fLab(y);
  const fz = fLab(z);
  return [116 * fy - 16, 500 * (fx - fy), 200 * (fy - fz)];
}

function fLab(t: number): number {
  return t > 0.008856 ? Math.cbrt(t) : 7.787 * t + 16 / 116;
}

/**
 * Nearest-palette lookup. For palettes above a size threshold a uniform-grid
 * LUT over sRGB cube is precomputed (32³ cells), so per-pixel lookup is O(1)
 * instead of O(paletteSize). The LUT is cached per (palette, metric) pair and
 * invalidated with the palette array identity.
 */
export interface PaletteLookup {
  readonly colors: PaletteColor[];
  readonly metric: ColorMetric;
  /** Nearest palette color for an sRGB byte triple. */
  find(r: number, g: number, b: number): PaletteColor;
}

const LUT_CACHE = new WeakMap<object, Map<ColorMetric, { lut: Uint16Array }>>();
const LUT_THRESHOLD = 24;

export function buildPaletteLookup(colors: PaletteColor[], metric: ColorMetric): PaletteLookup {
  if (colors.length === 0) {
    return { colors, metric, find: () => [0, 0, 0] };
  }
  if (colors.length < LUT_THRESHOLD) {
    return { colors, metric, find: (r, g, b) => nearestBrute(colors, metric, r, g, b) };
  }
  let byMetric = LUT_CACHE.get(colors);
  if (!byMetric) {
    byMetric = new Map();
    LUT_CACHE.set(colors, byMetric);
  }
  const cached = byMetric.get(metric);
  if (cached) {
    return { colors, metric, find: (r, g, b) => colors[cached.lut[gridIndex(r, g, b)]!]! };
  }
  const lut = new Uint16Array(LUT_GRID * LUT_GRID * LUT_GRID);
  const step = 256 / LUT_GRID;
  for (let ri = 0; ri < LUT_GRID; ri += 1) {
    for (let gi = 0; gi < LUT_GRID; gi += 1) {
      for (let bi = 0; bi < LUT_GRID; bi += 1) {
        const r = Math.round(ri * step + step / 2);
        const g = Math.round(gi * step + step / 2);
        const b = Math.round(bi * step + step / 2);
        const idx = (ri * LUT_GRID + gi) * LUT_GRID + bi;
        const [pr, pg, pb] = nearestBrute(colors, metric, r, g, b);
        lut[idx] = findIndexOf(colors, pr, pg, pb);
      }
    }
  }
  byMetric.set(metric, { lut });
  return { colors, metric, find: (r, g, b) => colors[lut[gridIndex(r, g, b)]!]! };
}

function gridIndex(r: number, g: number, b: number): number {
  const ri = Math.min(LUT_GRID - 1, Math.floor(r / (256 / LUT_GRID)));
  const gi = Math.min(LUT_GRID - 1, Math.floor(g / (256 / LUT_GRID)));
  const bi = Math.min(LUT_GRID - 1, Math.floor(b / (256 / LUT_GRID)));
  return (ri * LUT_GRID + gi) * LUT_GRID + bi;
}

function findIndexOf(colors: PaletteColor[], r: number, g: number, b: number): number {
  for (let i = 0; i < colors.length; i += 1) {
    const c = colors[i]!;
    if (c[0] === r && c[1] === g && c[2] === b) return i;
  }
  return 0;
}

function nearestBrute(
  colors: PaletteColor[],
  metric: ColorMetric,
  r: number,
  g: number,
  b: number,
): PaletteColor {
  let best = colors[0]!;
  let bestD = Infinity;
  for (let i = 0; i < colors.length; i += 1) {
    const c = colors[i]!;
    const d = paletteDistance(r, g, b, c[0], c[1], c[2], metric);
    if (d < bestD) {
      bestD = d;
      best = c;
    }
  }
  return best;
}

// ── Palette generation ─────────────────────────────────────────────────────

/**
 * Generate a palette from image data via median cut, optionally refined with
 * k-means iterations. Deterministic given (imageData, colorCount, metric,
 * seed): median cut never branches on random numbers and k-means seeds from
 * the deterministic PRNG.
 */
export function generatePalette(
  pixels: Uint8ClampedArray,
  w: number,
  h: number,
  colorCount: number,
  metric: ColorMetric,
  seed: number,
): PaletteColor[] {
  const count = Math.max(1, Math.min(256, Math.round(colorCount)));
  const sample = samplePixels(pixels, w, h);
  if (sample.length === 0) return [[0, 0, 0]];
  const box: Box = { pixels: sample, rMin: 0, rMax: 255, gMin: 0, gMax: 255, bMin: 0, bMax: 255 };
  const leaves = medianCut(box, count);
  let palette = leaves.map((leaf) => {
    let r = 0;
    let g = 0;
    let b = 0;
    for (const [pr, pg, pb] of leaf.pixels) {
      r += pr;
      g += pg;
      b += pb;
    }
    const n = leaf.pixels.length || 1;
    return [Math.round(r / n), Math.round(g / n), Math.round(b / n)] as PaletteColor;
  });
  // K-means refinement with deterministic centroids from median cut.
  palette = kmeansRefine(palette, sample, 6, metric, seed);
  return palette;
}

interface Box {
  pixels: PaletteColor[];
  rMin: number;
  rMax: number;
  gMin: number;
  gMax: number;
  bMin: number;
  bMax: number;
}

function samplePixels(pixels: Uint8ClampedArray, w: number, h: number): PaletteColor[] {
  const out: PaletteColor[] = [];
  const step = Math.max(1, Math.round(Math.sqrt((w * h) / 262144)));
  for (let y = 0; y < h; y += step) {
    for (let x = 0; x < w; x += step) {
      const o = (y * w + x) * 4;
      if (pixels[o + 3]! < 128) continue;
      out.push([pixels[o]!, pixels[o + 1]!, pixels[o + 2]!]);
    }
  }
  return out;
}

function medianCut(box: Box, target: number): Box[] {
  let boxes: Box[] = [box];
  while (boxes.length < target) {
    let largest = -1;
    let largestVolume = -1;
    for (let i = 0; i < boxes.length; i += 1) {
      const vol =
        (boxes[i]!.rMax - boxes[i]!.rMin + 1) *
        (boxes[i]!.gMax - boxes[i]!.gMin + 1) *
        (boxes[i]!.bMax - boxes[i]!.bMin + 1);
      if (vol > largestVolume) {
        largestVolume = vol;
        largest = i;
      }
    }
    if (largest < 0 || boxes[largest]!.pixels.length < 2) break;
    const split = splitBox(boxes[largest]!);
    if (!split) break;
    boxes = [...boxes.slice(0, largest), split[0], split[1], ...boxes.slice(largest + 1)];
  }
  return boxes;
}

function splitBox(box: Box): [Box, Box] | null {
  const rRange = box.rMax - box.rMin;
  const gRange = box.gMax - box.gMin;
  const bRange = box.bMax - box.bMin;
  let channel: 'r' | 'g' | 'b' = 'r';
  if (gRange >= rRange && gRange >= bRange) channel = 'g';
  else if (bRange >= rRange && bRange >= gRange) channel = 'b';
  const sorted = [...box.pixels].sort(
    (a, b) =>
      a[channel === 'r' ? 0 : channel === 'g' ? 1 : 2] -
      b[channel === 'r' ? 0 : channel === 'g' ? 1 : 2],
  );
  const mid = Math.floor(sorted.length / 2);
  if (mid === 0) return null;
  const lo = sorted.slice(0, mid);
  const hi = sorted.slice(mid);
  const makeBox = (pixels: PaletteColor[]): Box => {
    let rMin = 255;
    let rMax = 0;
    let gMin = 255;
    let gMax = 0;
    let bMin = 255;
    let bMax = 0;
    for (const [r, g, b] of pixels) {
      if (r < rMin) rMin = r;
      if (r > rMax) rMax = r;
      if (g < gMin) gMin = g;
      if (g > gMax) gMax = g;
      if (b < bMin) bMin = b;
      if (b > bMax) bMax = b;
    }
    return { pixels, rMin, rMax, gMin, gMax, bMin, bMax };
  };
  return [makeBox(lo), makeBox(hi)];
}

function kmeansRefine(
  centroids: PaletteColor[],
  samples: PaletteColor[],
  iterations: number,
  metric: ColorMetric,
  seed: number,
): PaletteColor[] {
  let centers = [...centroids];
  const rng = seed;
  for (let iter = 0; iter < iterations; iter += 1) {
    const sums = centers.map(() => [0, 0, 0, 0] as [number, number, number, number]);
    for (const [r, g, b] of samples) {
      let bestI = 0;
      let bestD = Infinity;
      for (let i = 0; i < centers.length; i += 1) {
        const c = centers[i]!;
        const d = paletteDistance(r, g, b, c[0], c[1], c[2], metric);
        if (d < bestD) {
          bestD = d;
          bestI = i;
        }
      }
      const s = sums[bestI]!;
      s[0] += r;
      s[1] += g;
      s[2] += b;
      s[3] += 1;
    }
    const next = centers.map((_c, i) => {
      const s = sums[i]!;
      if (s[3] === 0) {
        const jitter = Math.round(seeded01(rng + iter * 7919) * 255);
        return [jitter, jitter, jitter] as PaletteColor;
      }
      return [
        Math.round(s[0] / s[3]),
        Math.round(s[1] / s[3]),
        Math.round(s[2] / s[3]),
      ] as PaletteColor;
    });
    let changed = false;
    for (let i = 0; i < next.length; i += 1) {
      const a = next[i]!;
      const b = centers[i]!;
      if (a[0] !== b[0] || a[1] !== b[1] || a[2] !== b[2]) changed = true;
    }
    centers = next;
    if (!changed) break;
  }
  return centers;
}

/** Deduplicate palette colors (exact equality). */
export function dedupePalette(colors: PaletteColor[]): PaletteColor[] {
  const seen = new Set<number>();
  const out: PaletteColor[] = [];
  for (const c of colors) {
    const key = (c[0] << 16) | (c[1] << 8) | c[2];
    if (seen.has(key)) continue;
    seen.add(key);
    out.push(c);
  }
  return out;
}

/** Validate a palette from user input; drops malformed entries, caps size. */
export function sanitizePalette(colors: readonly (readonly number[])[]): PaletteColor[] {
  const out: PaletteColor[] = [];
  for (const c of colors) {
    if (!c || c.length < 3) continue;
    const r = Math.max(0, Math.min(255, Math.round(Number(c[0]) || 0)));
    const g = Math.max(0, Math.min(255, Math.round(Number(c[1]) || 0)));
    const b = Math.max(0, Math.min(255, Math.round(Number(c[2]) || 0)));
    out.push([r, g, b]);
    if (out.length >= 256) break;
  }
  return dedupePalette(out);
}

File: crates/varve-effects/src/palette_core.rs
//! Shared quantization core for the Dither and Palette Snap kernels —
//! faithful port of `packages/engine/src/liveEffects/paletteCore.ts`.

use crate::prng::{seeded01, srgb_to_linear01};
use serde::{Deserialize, Serialize};

pub type PaletteColor = [u8; 3];

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ColorMetric {
    Rgb,
    LinearRgb,
    Lab,
    Oklab,
}

impl ColorMetric {
    pub fn parse(s: &str) -> ColorMetric {
        match s {
            "linear-rgb" => ColorMetric::LinearRgb,
            "lab" => ColorMetric::Lab,
            "oklab" => ColorMetric::Oklab,
            _ => ColorMetric::Rgb,
        }
    }
}

const LUT_GRID: usize = 32;
const LUT_THRESHOLD: usize = 24;

/// Linear sRGB [r,g,b] (0-1) → Oklab [L, a, b] — port of
/// `linearSrgbToOklab` from @varve/shared colorConversion.ts.
pub fn linear_srgb_to_oklab(rgb: [f64; 3]) -> [f64; 3] {
    const M1: [[f64; 3]; 3] = [
        [0.4122214708, 0.5363325363, 0.0514459929],
        [0.2119034982, 0.6806995451, 0.1073969566],
        [0.0883024619, 0.2817188376, 0.6299787005],
    ];
    const M2: [[f64; 3]; 3] = [
        [0.2104542553, 0.793617785, -0.0040720468],
        [1.9779984951, -2.428592205, 0.4505937099],
        [0.0259040371, 0.7827717662, -0.808675766],
    ];
    let mut lms = [0.0f64; 3];
    for row in 0..3 {
        lms[row] = M1[row][0] * rgb[0] + M1[row][1] * rgb[1] + M1[row][2] * rgb[2];
    }
    let lms_cb: [f64; 3] = [lms[0].cbrt(), lms[1].cbrt(), lms[2].cbrt()];
    let mut out = [0.0f64; 3];
    for row in 0..3 {
        out[row] = M2[row][0] * lms_cb[0] + M2[row][1] * lms_cb[1] + M2[row][2] * lms_cb[2];
    }
    out
}

fn to_lab_space(r: f64, g: f64, b: f64, metric: ColorMetric) -> [f64; 3] {
    if metric == ColorMetric::Oklab {
        return linear_srgb_to_oklab([srgb_to_linear01(r), srgb_to_linear01(g), srgb_to_linear01(b)]);
    }
    // Lab via XYZ D65 (matches @varve/shared analytical conversion path).
    let rl = srgb_to_linear01(r);
    let gl = srgb_to_linear01(g);
    let bl = srgb_to_linear01(b);
    let mut x = rl * 0.4124 + gl * 0.3576 + bl * 0.1805;
    let y = rl * 0.2126 + gl * 0.7152 + bl * 0.0722;
    let mut z = rl * 0.0193 + gl * 0.1192 + bl * 0.9505;
    x /= 0.95047;
    z /= 1.08883;
    let fx = f_lab(x);
    let fy = f_lab(y);
    let fz = f_lab(z);
    [116.0 * fy - 16.0, 500.0 * (fx - fy), 200.0 * (fy - fz)]
}

fn f_lab(t: f64) -> f64 {
    if t > 0.008856 {
        t.cbrt()
    } else {
        7.787 * t + 16.0 / 116.0
    }
}

/// Squared distance between an sRGB byte colour and a palette colour —
/// port of `paletteDistance`.
pub fn palette_distance(r: f64, g: f64, b: f64, pr: f64, pg: f64, pb: f64, metric: ColorMetric) -> f64 {
    match metric {
        ColorMetric::Rgb => {
            let dr = r - pr;
            let dg = g - pg;
            let db = b - pb;
            dr * dr + dg * dg + db * db
        }
        ColorMetric::LinearRgb => {
            let lr = srgb_to_linear01(r) - srgb_to_linear01(pr);
            let lg = srgb_to_linear01(g) - srgb_to_linear01(pg);
            let lb = srgb_to_linear01(b) - srgb_to_linear01(pb);
            lr * lr + lg * lg + lb * lb
        }
        ColorMetric::Lab | ColorMetric::Oklab => {
            let [l1, a1, b1] = to_lab_space(r, g, b, metric);
            let [l2, a2, b2] = to_lab_space(pr, pg, pb, metric);
            let dl = l1 - l2;
            let da = a1 - a2;
            let db = b1 - b2;
            dl * dl + da * da + db * db
        }
    }
}

fn nearest_brute(colors: &[PaletteColor], metric: ColorMetric, r: f64, g: f64, b: f64) -> PaletteColor {
    let mut best = colors[0];
    let mut best_d = f64::INFINITY;
    for c in colors {
        let d = palette_distance(r, g, b, c[0] as f64, c[1] as f64, c[2] as f64, metric);
        if d < best_d {
            best_d = d;
            best = *c;
        }
    }
    best
}

fn find_index_of(colors: &[PaletteColor], r: u8, g: u8, b: u8) -> u16 {
    for (i, c) in colors.iter().enumerate() {
        if c[0] == r && c[1] == g && c[2] == b {
            return i as u16;
        }
    }
    0
}

fn grid_index(r: f64, g: f64, b: f64) -> usize {
    let ri = ((r / (256.0 / LUT_GRID as f64)).floor() as usize).min(LUT_GRID - 1);
    let gi = ((g / (256.0 / LUT_GRID as f64)).floor() as usize).min(LUT_GRID - 1);
    let bi = ((b / (256.0 / LUT_GRID as f64)).floor() as usize).min(LUT_GRID - 1);
    (ri * LUT_GRID + gi) * LUT_GRID + bi
}

/// Nearest-palette lookup with uniform-grid LUT acceleration for large
/// palettes — port of `buildPaletteLookup`. The LUT is built once per call
/// (the TS WeakMap cache keyed on array identity is not portable; callers
/// that reuse a palette should hold the returned struct).
pub struct PaletteLookup {
    colors: Vec<PaletteColor>,
    metric: ColorMetric,
    lut: Option<Vec<u16>>,
}

impl PaletteLookup {
    pub fn build(colors: &[PaletteColor], metric: ColorMetric) -> PaletteLookup {
        if colors.is_empty() {
            return PaletteLookup { colors: vec![], metric, lut: None };
        }
        if colors.len() < LUT_THRESHOLD {
            return PaletteLookup { colors: colors.to_vec(), metric, lut: None };
        }
        let mut lut = vec![0u16; LUT_GRID * LUT_GRID * LUT_GRID];
        let step = 256.0 / LUT_GRID as f64;
        for ri in 0..LUT_GRID {
            for gi in 0..LUT_GRID {
                for bi in 0..LUT_GRID {
                    let r = (ri as f64 * step + step / 2.0).round();
                    let g = (gi as f64 * step + step / 2.0).round();
                    let b = (bi as f64 * step + step / 2.0).round();
                    let idx = (ri * LUT_GRID + gi) * LUT_GRID + bi;
                    let [pr, pg, pb] = nearest_brute(colors, metric, r, g, b);
                    lut[idx] = find_index_of(colors, pr, pg, pb);
                }
            }
        }
        PaletteLookup { colors: colors.to_vec(), metric, lut: Some(lut) }
    }

    /// Nearest palette color for an sRGB byte triple.
    pub fn find(&self, r: u8, g: u8, b: u8) -> PaletteColor {
        if self.colors.is_empty() {
            return [0, 0, 0];
        }
        match &self.lut {
            Some(lut) => {
                let idx = grid_index(r as f64, g as f64, b as f64);
                self.colors[lut[idx] as usize]
            }
            None => nearest_brute(&self.colors, self.metric, r as f64, g as f64, b as f64),
        }
    }
}

// ── Palette generation ─────────────────────────────────────────────────────

#[derive(Clone)]
struct Box {
    pixels: Vec<PaletteColor>,
    r_min: u8,
    r_max: u8,
    g_min: u8,
    g_max: u8,
    b_min: u8,
    b_max: u8,
}

/// Generate a palette via median cut + k-means refinement — port of
/// `generatePalette`.
pub fn generate_palette(
    pixels: &[u8],
    w: u32,
    h: u32,
    color_count: f64,
    metric: ColorMetric,
    seed: u32,
) -> Vec<PaletteColor> {
    let count = (color_count.round()).max(1.0).min(256.0) as usize;
    let sample = sample_pixels(pixels, w, h);
    if sample.is_empty() {
        return vec![[0, 0, 0]];
    }
    let box0 = Box {
        pixels: sample.clone(),
        r_min: 0,
        r_max: 255,
        g_min: 0,
        g_max: 255,
        b_min: 0,
        b_max: 255,
    };
    let leaves = median_cut(box0, count);
    let mut palette: Vec<PaletteColor> = leaves
        .iter()
        .map(|leaf| {
            let mut r = 0u64;
            let mut g = 0u64;
            let mut b = 0u64;
            for [pr, pg, pb] in &leaf.pixels {
                r += *pr as u64;
                g += *pg as u64;
                b += *pb as u64;
            }
            let n = if leaf.pixels.is_empty() { 1 } else { leaf.pixels.len() as u64 };
            [
                (r as f64 / n as f64).round() as u8,
                (g as f64 / n as f64).round() as u8,
                (b as f64 / n as f64).round() as u8,
            ]
        })
        .collect();
    palette = kmeans_refine(palette, &sample, 6, metric, seed);
    palette
}

fn sample_pixels(pixels: &[u8], w: u32, h: u32) -> Vec<PaletteColor> {
    let mut out = Vec::new();
    let step = (((w * h) as f64 / 262144.0).sqrt().round()).max(1.0) as u32;
    let mut y = 0;
    while y < h {
        let mut x = 0;
        while x < w {
            let o = ((y * w + x) * 4) as usize;
            if pixels[o + 3] >= 128 {
                out.push([pixels[o], pixels[o + 1], pixels[o + 2]]);
            }
            x += step;
        }
        y += step;
    }
    out
}

fn median_cut(box0: Box, target: usize) -> Vec<Box> {
    let mut boxes = vec![box0];
    while boxes.len() < target {
        let mut largest = -1i64;
        let mut largest_volume = -1i64;
        for (i, b) in boxes.iter().enumerate() {
            let vol = (b.r_max as i64 - b.r_min as i64 + 1)
                * (b.g_max as i64 - b.g_min as i64 + 1)
                * (b.b_max as i64 - b.b_min as i64 + 1);
            if vol > largest_volume {
                largest_volume = vol;
                largest = i as i64;
            }
        }
        if largest < 0 || boxes[largest as usize].pixels.len() < 2 {
            break;
        }
        let split = split_box(boxes[largest as usize].clone());
        let Some((lo, hi)) = split else {
            break;
        };
        let mut next = Vec::with_capacity(boxes.len() + 1);
        next.extend_from_slice(&boxes[..largest as usize]);
        next.push(lo);
        next.push(hi);
        next.extend_from_slice(&boxes[largest as usize + 1..]);
        boxes = next;
    }
    boxes
}

fn split_box(box0: Box) -> Option<(Box, Box)> {
    let r_range = box0.r_max as i32 - box0.r_min as i32;
    let g_range = box0.g_max as i32 - box0.g_min as i32;
    let b_range = box0.b_max as i32 - box0.b_min as i32;
    let channel: u8 = if g_range >= r_range && g_range >= b_range {
        1
    } else if b_range >= r_range && b_range >= g_range {
        2
    } else {
        0
    };
    let mut sorted = box0.pixels.clone();
    sorted.sort_by_key(|c| c[channel as usize]);
    let mid = sorted.len() / 2;
    if mid == 0 {
        return None;
    }
    let lo = sorted[..mid].to_vec();
    let hi = sorted[mid..].to_vec();
    Some((make_box(lo), make_box(hi)))
}

fn make_box(pixels: Vec<PaletteColor>) -> Box {
    let mut r_min = 255u8;
    let mut r_max = 0u8;
    let mut g_min = 255u8;
    let mut g_max = 0u8;
    let mut b_min = 255u8;
    let mut b_max = 0u8;
    for [r, g, b] in &pixels {
        if *r < r_min {
            r_min = *r;
        }
        if *r > r_max {
            r_max = *r;
        }
        if *g < g_min {
            g_min = *g;
        }
        if *g > g_max {
            g_max = *g;
        }
        if *b < b_min {
            b_min = *b;
        }
        if *b > b_max {
            b_max = *b;
        }
    }
    Box { pixels, r_min, r_max, g_min, g_max, b_min, b_max }
}

fn kmeans_refine(
    centroids: Vec<PaletteColor>,
    samples: &[PaletteColor],
    iterations: usize,
    metric: ColorMetric,
    seed: u32,
) -> Vec<PaletteColor> {
    let mut centers = centroids;
    for iter in 0..iterations {
        let mut sums = vec![[0u64; 4]; centers.len()];
        for [r, g, b] in samples {
            let mut best_i = 0usize;
            let mut best_d = f64::INFINITY;
            for (i, c) in centers.iter().enumerate() {
                let d = palette_distance(
                    *r as f64,
                    *g as f64,
                    *b as f64,
                    c[0] as f64,
                    c[1] as f64,
                    c[2] as f64,
                    metric,
                );
                if d < best_d {
                    best_d = d;
                    best_i = i;
                }
            }
            let s = &mut sums[best_i];
            s[0] += *r as u64;
            s[1] += *g as u64;
            s[2] += *b as u64;
            s[3] += 1;
        }
        let next: Vec<PaletteColor> = centers
            .iter()
            .enumerate()
            .map(|(i, _c)| {
                let s = sums[i];
                if s[3] == 0 {
                    let jitter = (seeded01(seed.wrapping_add((iter * 7919) as u32)) * 255.0).round() as u8;
                    [jitter, jitter, jitter]
                } else {
                    [
                        (s[0] as f64 / s[3] as f64).round() as u8,
                        (s[1] as f64 / s[3] as f64).round() as u8,
                        (s[2] as f64 / s[3] as f64).round() as u8,
                    ]
                }
            })
            .collect();
        let changed = next.iter().zip(centers.iter()).any(|(a, b)| a != b);
        centers = next;
        if !changed {
            break;
        }
    }
    centers
}

/// Deduplicate palette colors (exact equality) — port of `dedupePalette`.
pub fn dedupe_palette(colors: &[PaletteColor]) -> Vec<PaletteColor> {
    let mut seen = std::collections::HashSet::new();
    let mut out = Vec::new();
    for c in colors {
        let key = ((c[0] as u32) << 16) | ((c[1] as u32) << 8) | c[2] as u32;
        if seen.insert(key) {
            out.push(*c);
        }
    }
    out
}

/// Validate a palette from user input; drops malformed entries, caps size —
/// port of `sanitizePalette`.
pub fn sanitize_palette(colors: &[[f64; 3]]) -> Vec<PaletteColor> {
    let mut out = Vec::new();
    for c in colors {
        let r = (c[0].max(0.0).min(255.0).round()) as u8;
        let g = (c[1].max(0.0).min(255.0).round()) as u8;
        let b = (c[2].max(0.0).min(255.0).round()) as u8;
        out.push([r, g, b]);
        if out.len() >= 256 {
            break;
        }
    }
    dedupe_palette(&out)
}
